File: python/example9_4.py
# Example 9.4 Multiplicative Heteroscedasticity
# Multiplicative Heteroscedasticity

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf

data = pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/greene.txt',sep='\s+',nrows=51)
data = data.dropna()  # take care of missing obs
print(data.describe())

spending = data['SPENDING']
income = data['INCOME']/10000

df1 = pd.concat([spending,income],keys=['spending','income'],axis=1)
model0 = smf.ols(formula='spending~income+I(income**2)',data = df1).fit()
b0 = model0.params

# log likelihood function
from numpy import sqrt, mean, log, exp
from scipy import stats, optimize

def llf1obs(b):
    h = income**b[3]
    e = (spending-b[0]-b[1]*income-b[2]*income**2)/sqrt(h)
    ll = log(stats.norm.pdf(e, loc=0, scale=sqrt(mean(e**2))))
    return (ll-0.5*log(h))

def llf2obs(b):
    h = exp(income*b[3])
    e = (spending-b[0]-b[1]*income-b[2]*income**2)/sqrt(h)
    ll = log(stats.norm.pdf(e, loc=0, scale=sqrt(mean(e**2))))
    return (ll-0.5*log(h))

# Maximum Likelihood using statsmodels' GenericLikelihoodModel
from statsmodels.base.model import GenericLikelihoodModel

class llf_het1(GenericLikelihoodModel):
    def loglikeobs(self, params):
        b = params
        return llf1obs(b)

class llf_het2(GenericLikelihoodModel):
    def loglikeobs(self, params):
        b = params
        return llf2obs(b)

b = list(b0) + [0]  # initial values
    
het1_model = llf_het1(spending,income)  # data input is a dummy
het1_results = het1_model.fit(b,method='newton')
print(het1_results.summary())

het2_model = llf_het2(spending,income)  # data input is a dummy
het2_results = het2_model.fit(b,method='newton')
print(het2_results.summary())

File: R/example9_4.R
# Example 9.4 Multiplicative Heteroscedasticity
# Multiplicative Heteroscedasticity

setwd("C:/Course17/ceR/R")
greene<-read.table("http://web.pdx.edu/~crkl/ceR/data/greene.txt",header=T,nrows=51)
# file name may be case sensitive!

greene<-na.omit(greene)  # take care of missing obs
summary(greene)

x<-greene[,3]/10000  # income
y<-greene[,2]        # spending

model0<-lm(y~x+I(x^2))
b0<-model0$coefficients

# log-likelihood function (data: y,x)
llf1<-function(b) {
  h<-x^b[4]           # multiplicative hetero.
  # h<-exp(x*b[4])
  e<-(y-b[1]-b[2]*x-b[3]*(x^2))/sqrt(h)
  ll<-dnorm(e,mean=0,sd=sqrt(mean(e^2)),log=T)
  ll-0.5*log(h)
}

llf2<-function(b) {
  # h<-x^b[4]           # multiplicative hetero.
  h<-exp(x*b[4])
  e<-(y-b[1]-b[2]*x-b[3]*(x^2))/sqrt(h)
  ll<-dnorm(e,mean=0,sd=sqrt(mean(e^2)),log=T)
  ll-0.5*log(h)
}

# Maximum Likelihood
library(maxLik)   # need to install maxLik package
model1<-maxLik(llf1,start=c(b0,0),method="BHHH")
summary(model1)
model2<-maxLik(llf2,start=c(b0,0),method="BHHH")
summary(model2)

rm(list=ls())
